ra/ra.go

```diff
@@ -1846,57 +2156,94 @@ func (ra *RegistrationAuthorityImpl) AdministrativelyRevokeCertificate(ctx conte
-		return nil, fmt.Errorf("cannot revoke for KeyCompromise by serial alone")
+	reasonCode := revocation.Reason(req.Code)
+	if reasonCode == ocsp.KeyCompromise && req.Cert == nil && !req.SkipBlockKey {
+		return nil, fmt.Errorf("cannot revoke and block for KeyCompromise by serial alone")
 	}
-	if req.SkipBlockKey && revocationCode != ocsp.KeyCompromise {
+	if req.SkipBlockKey && reasonCode != ocsp.KeyCompromise {
 		return nil, fmt.Errorf("cannot skip key blocking for reasons other than KeyCompromise")
 	}
 
+	if _, present := revocation.AdminAllowedReasons[reasonCode]; !present {
+		return nil, fmt.Errorf("cannot revoke for reason %d", reasonCode)
+	}
+
+	// If we don't have a real cert, we create a fake cert (containing just the
+	// serial number, which is all we need) and look up the IssuerID from the db.
+	// We could instead look up and parse the certificate itself, but we avoid
+	// that in case we are administratively revoking the certificate because it is
+	// so badly malformed that it can't be parsed.
 	var cert *x509.Certificate
-	var serialString string
+	var issuerID int64 // TODO(#5152) make this an issuance.IssuerNameID
 	var err error
-	if req.Cert != nil {
-		cert, err = x509.ParseCertificate(req.Cert)
+	if req.Cert == nil {
+		cert = nil
+		serial, err := core.StringToSerial(req.Serial)
 		if err != nil {
 			return nil, err
 		}
-		serialString = core.SerialToString(cert.SerialNumber)
+		cert = &x509.Certificate{
+			SerialNumber: serial,
+		}
+
+		status, err := ra.SA.GetCertificateStatus(ctx, &sapb.Serial{Serial: req.Serial})
+		if err != nil {
+			return nil, fmt.Errorf("unable to confirm that serial %q was ever issued: %w", serial, err)
+		}
+		issuerID = status.IssuerID
 	} else {
-		serialNum, err := core.StringToSerial(req.Serial)
+		cert, err = x509.ParseCertificate(req.Cert)
 		if err != nil {
 			return nil, err
 		}
-		cert = &x509.Certificate{
-			SerialNumber: serialNum,
-		}
-		serialString = req.Serial
+		issuerID = int64(issuance.GetIssuerNameID(cert))
 	}
 
-	state := "Failure"
+	logEvent := certificateRevocationEvent{
+		ID:           core.NewToken(),
+		Method:       "key",
+		AdminName:    req.AdminName,
+		SerialNumber: core.SerialToString(cert.SerialNumber),
+	}
 	defer func() {
-		// Needed:
-		//   Serial
-		//   CN
-		//   DNS names
-		//   Revocation reason
-		//   Name of admin-revoker user
-		//   Error (if there was one)
-		ra.log.AuditInfof(
-			"%s, admin-revoker user: %s",
-			revokeEvent(state, serialString, cert.Subject.CommonName, cert.DNSNames, revocationCode),
-			req.AdminName)
+		if err != nil {
+			logEvent.Error = err.Error()
+		}
+		ra.log.AuditObject("Revocation request:", logEvent)
 	}()
 
-	err = ra.revokeCertificate(ctx, cert, revocationCode, 0, "admin-revoker", fmt.Sprintf("revoked by %s", req.AdminName), req.SkipBlockKey)
+	err = ra.revokeCertificate(ctx, cert.SerialNumber, issuerID, revocation.Reason(req.Code))
+	if err != nil {
```

Comment:
If we're not using a sentinel error here we should add a comment line for why we're doing this.

Reply:
Using the same semi-sentinel error as for the comment above.